### DISCUSSION TITLE
Program is not generating intended action

### DISCUSSION THREAD

**Discussion by DJXer** [pybricks/support#1701] [Q&A]
Hello,

I just purchased the lifetime license for Pybricks earlier today, and decided to use the block-based method as an alternative to Lego's Powered Up app. I decided to make a little test program to start with, however I'm running into an unexpected problem that is worrying me about whether or not I will be able to use this.

My intended action is the following: _Using a technic hub, make a small PUP motor (with rotation sensor) run continuously._

Below I will attach a screenshot of the code I have.

<img width="579" alt="Screenshot 2024-07-04 at 8 05 36 PM" src="https://github.com/pybricks/support/assets/174753430/704b8d35-c6b2-4ce6-ba84-2f2b753b47be">

I watched a video ( https://www.youtube.com/watch?v=wZc1ix00cZY ) online showing a similar layout to compare. It didn't use the exact same motors/hubs, but conceptually it was the same set up and program. Below are some more details:

- My motor is indeed one of the motors listed under the "motors with rotation sensors" page
- Motor is indeed plugged into port A on my technic hub
- Pybricks firmware is installed onto the technic hub
- Hub is indeed connected via bluetooth to the computer
-  When the play button is pressed, the program seems to start and immediately end
- The block shown below _DOES_ seem to run properly:

<img width="421" alt="Screenshot 2024-07-04 at 8 19 02 PM" src="https://github.com/pybricks/support/assets/174753430/beef57a2-e278-4b7e-9d73-8d03044a0dd4">

- I tried to add the block that changes the light on the technic hub. This block too does not work by itself _OR_ with the "run motor forever" block, but the light _DOES_ seem to change when this block is used with the block in the above image.
- I have tried other motors of the same type and of different sizes using the original code. Problem still persists
- I have tried this with and without other motors plugged in. Problem persists
- I have connected and reconnected the technic hub, problem persists
- Round progress bar when the play button is pressed sometimes freezes. Seemingly the only solution is to disconnect and reconnect the hub
- Nothing shows up in the errors log

I have a good deal of experience with block-based coding, and I don't _believe_ there's a problem with the code I created. It's entirely possible though that I missed something crucial...I just have no idea what that could be, which is why I am asking here. Any and all help is appreciated, thanks for reading.

Edit — I should also mention that I am _NOT_ using batteries in my technic hub, but it is instead receiving direct power. Not sure if this would matter but I thought it would be good to mention anyway.

Edit 2 — Haha...Very interesting, but I figured it out. I needed a "wait forever" block at the end of all the code. Is this the correct solution, or should it also work without adding that? I'll leave this open so that anyone can reply if they have thoughts. One suggestion I have though is to make a note of this in the block information pages, as this isn't listed there.

**Comment by laurensvalk** [comment]
Thank you for supporting the project by purchasing a license for block coding! This is a big help.

I'm glad you got it to work as intended.

> I figured it out. I needed a "wait forever" block at the end of all the code. Is this the correct solution, or should it also work without adding that?

This is correct. When all blocks in a program complete, it ends and the motors/sounds/lights stop. As you'll build larger programs in the future, you might find that this is quite practical for most situations. This way, your motors won't accidentally run your robot off your desk unless you intend it to :smile: 

So if the block that starts running the motor is the only thing in the program, it completes right away and you see no movement. There is a hint in the description of "forever", but we're also adding this to the upcoming, more detailed starter guide.

![image](https://github.com/pybricks/support/assets/12326241/ebdbb4e7-e798-49d8-bc4b-017e2e067895)

You've already found the solution for your use case, but I'll just add it here in case others come across this thread.

![image](https://github.com/pybricks/support/assets/12326241/8bef9b3d-23b4-40b1-acf0-50b3c559a155)

Thanks again!
